### ISSUE TITLE
no-constant-condition should allow for while(true) idiom

### ISSUE THREAD

**Issue by aparajita** [eslint/eslint#5477]
**What version of ESLint are you using?**

2.3.0

**What configuration and parser (Espree, Babel-ESLint, etc.) are you using?**

env: es6 + node
parser: Espree
config: "no-constant-condition": "error"

**What did you do? Please include the actual source code causing the issue.**

```js
while (true)
{
    if (doSomething())
        break;
}
```

**What did you expect to happen?**

I expected no error at the `while`, because its dependent statement has a `break`, and in this case it's equivalent to `for(;;)`. Maybe I'm the only one who uses this idiom.

**What actually happened? Please include the actual, raw output from ESLint.**

`line 15, col 1, Error - Unexpected constant condition. (no-constant-condition)`

line 15, col 1 is the "w" in `while`.

**What do you suggest?**

There are three approaches to fixing this:

1. Ignore it and hope I go away.  :wink:
1. Always allow `while (true)` the way `for (;;)` is allowed.
1. Add an "allow-while-true" option.

I'll be happy to implement it, but let me know which approach you want to take.

**Comment by platinumazure** [comment]
Just out of curiosity, why not `for (;;)`?

**Comment by ilyavolodin** [comment]
If `for(;;)` is allowed, I would expect `while(true)` to act the same way. @mysticatea any reason why one is enabled and the other one is not?

**Comment by mysticatea** [comment]
Thank you for this issue.

Though I didn't write this rule, I guess `for (;;)` doesn't have the expression node of the condition, so this rule isn't warning it.
`for (;;)` has the constant condition, so it should be warned by this rule, IMO.

**Comment by ilyavolodin** [comment]
Ahh.. sorry, I thought you wrote this rule (you write too many rules, so you are a number one suspect!:-) So you don't agree with the idea of adding `while(true)` and instead want to remove `for (;;)` right?

**Comment by platinumazure** [comment]
I don't think we should remove `for (;;)`. Besides being a breaking change, there is no condition expression, so it can't be said to be a constant condition.

**Comment by mysticatea** [comment]
Hmm, [the document](http://eslint.org/docs/rules/no-constant-condition) says `for (;;)` is no problem. So it was intentional. And yes, adding the warning for `for (;;)` is a breaking change. 
(I still think `no-constant-condition` should handle `for (;true;)` and `for (;;)` by the same way in future)

I'm OK adding an option to allow `while (true)`.
It's one of well-known idioms.

**Comment by aparajita** [comment]
The ECMAScript spec is quite clear that if the loop condition is empty in a `for` loop, it returns `true`. So `for (;;)` is **exactly** equivalent to `for (; true;)`, which by a strict definition is a constant condition. But clearly this is a common idiom which is useful, explicitly written, and thus should be allowed.

In exactly the same way, `while (true)` should be allowed. For consistency, I would advocate two boolean options for this rule, "allow-for" and "allow-while", both of which default to true.

**Comment by michaelficarra** [comment]
I think the rule should by default only report conditionals (`if` and `?:`), since they are certainly errors, and have an option to also report loops (`for`, `while`, `do`-`while`). Either way, `for(;;);` must not be treated any differently than `for(;true;);`.

**Comment by aparajita** [comment]
@michaelficarra Right, "allow-do-while" should also be an option that defaults to true.

**Comment by michaelficarra** [comment]
@aparajita No, I am saying all loops should be enabled by one option. And it should default to disabled.

**Comment by aparajita** [comment]
@michaelficarra Just to be clear, you mean reporting all loops as errors should be enabled by one option? I agree, but typically ESLint rule options are phrased as "allow-", so I think it's more consistent (and easier to understand) if it's expressed as "allow-loops" and defaults to true. However, there is also something to be said for having separate "allow-for", "allow-while", and "allow-do-while" options, because a user might want to enforce a stylistic preference for `while (true)` over `for (;;)` for example.

**Comment by michaelficarra** [comment]
@aparajita That's a fair point, but maybe handling that granularity should be the responsibility of another rule.

**Comment by aparajita** [comment]
@michaelficarra Why another rule? There has been a lot of effort put into coalescing rules with overlapping responsibility.

**Comment by isiahmeadows** [comment]
I feel an extra option for `no-constant-condition` to allow explicit boolean literals would be nice. I personally find `while (true)` more readable and the intent more explicit than `for (;;)`. Also, there's the `do { ... } while (false)` idiom, for which there is no substitute other than labelled blocks, which IMHO are ugly.

```js
for (;;) {
  doSomething()
  if (condition()) return null
}

while (true) {
  doSomething()
  if (condition()) return null
}
```

As for boolean literal conditions in `if (true)`, etc., that can still be caught by `no-unreachable-code`.

So :+1: from me.

**Comment by nzakas** [comment]
I'd propose a `checkLoops` option with a default value of `true` (which matches the current default). When setting it to `false`, it would just not apply to `for`, `while`, `do-while`, etc.

Would that satisfy the request here?

**Comment by aparajita** [comment]
Fine, although it's kind of inconsistent to allow by default `for (;;)`, because that **is** a constant condition. But whatever, as long as we have some way to allow the other common idioms.

**Comment by isiahmeadows** [comment]
@nzakas That would be perfect!

@aparajita I do find it somewhat inconsistent that it isn't counted, but keep in mind that literally *doesn't* have a condition at all, so arguments could go either way.

**Comment by nzakas** [comment]
@aparajita @isiahmeadows do either of you want to submit a PR for this?

**Comment by aparajita** [comment]
@nzakas I'll see if I can find the time. I made a version for myself that allows while(true), but of course you have a million other requirements for a PR.

**Comment by BYK** [comment]
@aparajita we don't really have that much. We just want you to conform to a few writing standards mostly. Anything I (or anyone on the team) can help with? Feel free to come by our Gitter channel for any help too.

**Comment by aparajita** [comment]
Not much, just all of [this](http://eslint.org/docs/developer-guide/contributing/pull-requests). I've contributed before, I don't need help, I just need time.

**Comment by kaicataldo** [comment]
I can take a look at this if @aparajita isn't already working on it

**Comment by aparajita** [comment]
@kaicataldo Go for it.
